File: go/variables/variables.go
package main

import (
	"fmt"
)

func main() {
	/* Go provides two primary ways to declare variables. The first is using the
	 *	"var" keyword, followed by the variable's name and type. Variables
	 *	declared in this manner do not require a value to be assigned to them at
	 *	the time of creation:*/
	var emptyMsg string

	/* Note that Go technically requires all variables to have a value assigned
	 *	to them. Even if a variable declared in the above manner has no value
	 *	explicitly assigned to it, Go will implicitly assign a default value to
	 *	the variable:*/
	fmt.Println(emptyMsg)

	/* Unlike many other languages, Go does not allow any unused local
	 *	variables.*/

	/* Variables created with the var keyword can also be declared without an
	 *	explicit type. In this case, however, a value must be assigned to the
	 *	variable so that Go can infer its type:*/
	var secondMsg = "Hello, world!"

	fmt.Println(secondMsg)

	/* The second primary way in which Go allows variables to be declared is via
	 *	the := operator. Variables created via the := operator do not use a type
	 *	keyword. Their type is always inferred. As such, variables declared in
	 *	this manner require a value to be assigned to them at the time of
	 *	creation:*/
	thirdMsg := "Hello again, world!"

	fmt.Println(thirdMsg)

	/* Go also allows multiple variables to be declare on the same line. When
	 *	done using the var keyword, the variables must all be of the same
	 *	type:*/
	var a, b, c int
	a = 1
	b = 2
	c = 3

	fmt.Println(a, b, c)

	/* When variables created with the var keyword don't include a type,
	 *	the variables cajn be of different types:*/
	var d, e, f = "one", "two", 3

	fmt.Println(d, e, f)

	// Multiple variables can also be declared via the := operator:
	fourthMessage, fifthMessage := "one", 2

	fmt.Println(fourthMessage, fifthMessage)

	/* Go also allows variable declarations to be grouped into a block for
	 *	easier readability:*/
	var (
		g int
		h        = 1
		i string = "two"
	)

	fmt.Println(g, h, i)

	// Constants can be declared via the const keyword:
	const pi = 3.14

	/* As with variables, a constant's type will be inferred by the compiler if
	 * it isn't explicitly defined. All constants require that a value be
	 * assigned at the time of declaration. Unlike local variables, Go allows
	 * unused constants.*/

}

File: c/variables/variables.c
#include <stdio.h>

int main()
{
    /* A variable can be thought of as a container that holds a value. C is a
     *  strongly typed language, meaning all variables must be given a specific,
     *  defined type. In C, variables are declared by specifying the variable's
     *  type, followed by a name known as an identifer that can be used to refer
     *  to the variable.
     * 
     * The process of creating a variable is referred to as declaration. The
     *  statement below declares a variable with an identifier of x:*/
    int x;

    /* Identifiers must be unique to their scope, and cannot be the same as any
     *  of the language's reserved words. They must begin with either a letter
     *  (upper or lower case) or an underscore, and can only contain underscores
     *  and alphanumeric characters.*/

    // To assign a value to a variable (initialization), use the = operator:
    x = 15;
    printf("x: %i\n", x);

    // A variable can also be initialized at declaration:
    int y = 20;
    printf("y: %i\n", y);

    /* Multiple variables of the same type can also be delcared and initialized
     *  as part of a single statement. When this is done, the type is specified
     *  once at the beginning of the statement, and each variable declaration is
     *  separated by a comma:*/
    int a = 1,
        b,
        c = 3;
        
    b = 2;

    printf("a: %i\nb: %i\nc: %i\n", a, b, c);

    /* Constants are similar to variables. Unlike variables, however, their
     *  value cannot be changed after initialization. To declare a constant,
     *  the const keyword is used, followed by the constant's type and
     *  identifier:*/
    const int constant = 95;
    printf("constant: %i\n", constant);

    return 0;

}